Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,640 @@
+
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _mantissaAddition(NSDecimalInternal* result, NSDecimalInternal* augend, unsigned short addend) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // Add the addend to the mantissa
+    for (int index = 0; index < augend->decimal._length; ++index) {
+        unsigned int res = static_cast<unsigned int>((augend->_mantissa)[index]) + static_cast<unsigned int>(addend);
+        (result->_mantissa)[index] = res;
+        addend = res >> 0x10;
+    }
+
+    if (addend != 0) {
+        if (result->_extendedLen == augend->decimal._length) {
+            // overflow
+            return NSCalculationOverflow;
+        }
+
+        // set the carryover
+        (result->_mantissa)[augend->decimal._length] = addend;
+        result->decimal._length++;
+    }
+
+    return NSCalculationNoError;
+}
+
+static NSCalculationError _mantissaMultiplication(NSDecimalInternal* result, NSDecimalInternal* multiplicand, unsigned short multiplier) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    // This will be used for by hand style multiplication
+    if (multiplier == 0 || multiplicand->decimal._length == 0) {
+        result->decimal._length = 0;
+        result->decimal._exponent = 0;
+        return NSCalculationNoError;
+    }
+
+    unsigned int carryOver = 0;
+
+    // multiply the mantissa by the multiplier and propagate the carry over.
+    for (int index = 0; index < multiplicand->decimal._length; ++index) {
+        unsigned int res =
+            (static_cast<unsigned int>(multiplier) * static_cast<unsigned int>((multiplicand->_mantissa)[index])) + carryOver;
+        (result->_mantissa)[index] = res;
+        carryOver = res >> 0x10;
+    }
+
+    // we have the carry over
+    if (carryOver != 0) {
+        if (result->_extendedLen == multiplicand->decimal._length) {
+            // overflow
+            return NSCalculationOverflow;
+        }
+
+        // set the overflow
+        (result->_mantissa)[multiplicand->decimal._length] = carryOver;
+        result->decimal._length++;
+    }
+
+    return NSCalculationNoError;
+}
+
+static unsigned int _obtainLength(unsigned short* mantissa, int length) {
+    unsigned int result = 0;
+    for (int i = length - 1; i >= 0; --i) {
+        result = i + 1;
+        if (mantissa[i] != 0) {
+            break;
+        }
+    }
+
+    return result;
+}
+
+static NSCalculationError _mantissaDivision(NSDecimalInternal* result, NSDecimalInternal* dividend, unsigned short divisor) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    if (!divisor) {
+        INVALIDATE_DECIMAL((&result->decimal));
+        return NSCalculationDivideByZero;
+    }
+
+    // Divide the mantissa by the divisor
+    for (int index = dividend->decimal._length - 1; index >= 0; --index) {
+        unsigned int res = static_cast<unsigned int>((dividend->_mantissa)[index]) | (result->_remainder << 0x10);
```

Comment:
it's  a | since you are casting the unsigned short into an int and then pretty much adding onto the remainder.


---
In reply to: [81243253](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81243253) [](ancestors = 81243253)